### PR TITLE
[CD] Add Python Docker Pipeline

### PR DESCRIPTION
**Pull request by perdasilva**
* Refactors docker login functionality out of ci/docker_cache.py to make it generic
* Refactors docker run functionality out of ci/build.py to make it general purpose
* Updates CD pipeline for producing python docker images

### CODE REVIEW

File: cd/python/docker/python_images.sh [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        runtime_param="--runtime=nvidia"
+    fi
+    local test_image_name="${image_name}_test"
+    
+    # Ensure the correct context root is passed in when building - Dockerfile.test expects ci directory
+    docker build -t "${test_image_name}" --build-arg USER_ID=`id -u` --build-arg GROUP_ID=`id -g` --build-arg BASE_IMAGE="${image_name}" -f ${resources_path}/Dockerfile.test ./ci
+    ./ci/safe_docker_run.py ${runtime_param} --cap-add "SYS_PTRACE" -u `id -u`:`id -g` -v `pwd`:/work/mxnet "${test_image_name}" ${resources_path}/test_python_image.sh "${mxnet_variant}" "${python_cmd}"
+}
+
+# Pushes the runtime image to the repository
```
Comment (perdasilva) [inline cd/python/docker/python_images.sh]: Comments don't seem related to this docker image...

File: ci/docker_cache.py [hunk 2]
```diff
@@ -24,21 +24,21 @@
...
 import build as build_util
+from docker_login import login_dockerhub, logout_dockerhub
 from util import retry
 
-DOCKERHUB_LOGIN_NUM_RETRIES = 5
-DOCKERHUB_RETRY_SECONDS = 5
-DOCKER_CACHE_NUM_RETRIES = 3
 DOCKER_CACHE_TIMEOUT_MINS = 45
 PARALLEL_BUILDS = 10
+DOCKER_CACHE_NUM_RETRIES = 3
```
Comment (perdasilva) [inline ci/docker_cache.py]: Move this to the top to minimize change set

File: ci/docker_login.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    # > Configure a credential helper to remove this warning. See
+    # > https://docs.docker.com/engine/reference/commandline/login/#credentials-store
+    # Since we consider the restricted slaves a secure environment, that's fine. Also, using this will require
+    # third party applications which would need a review first as well.
+    p = subprocess.run(['docker', 'login', '--username', dockerhub_credentials['username'], '--password-stdin'],
+                       stdout=subprocess.PIPE, input=str.encode(dockerhub_credentials['password']))
+    logging.info(p.stdout)
+    if p.returncode == 0:
+        logging.info('Successfully logged in to DockerHub')
+        return True
```
Comment (perdasilva) [inline ci/docker_login.py]: No need to return true here - failure results in Error getting raised

File: cd/python/docker/Jenkins_pipeline.groovy [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    return "ubuntu_gpu_${mxnet_variant.replace('mkl', '')}"
+  }
+  return "ubuntu_cpu"
+}
+
+
+def build(mxnet_variant) {
+  ws("workspace/python_docker/${mxnet_variant}/${env.BUILD_NUMBER}") {
+    ci_utils.init_git()
+    cd_utils.restore_artifact(mxnet_variant, 'static')
```
Comment (perdasilva) [inline cd/python/docker/Jenkins_pipeline.groovy]: Maybe create two variants of this method one for static and one for dynamic, instead of passing this in as a parameter. Then, call the methods restore_static_libmxnet and restore_dynamic_libmxnet

File: ci/test_safe_docker_run.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        """
+        Tests 152 is returned if an error is raised when calling container.remove
+        """
+        mock_container = create_mock_container()
+        mock_container.remove.side_effect = RuntimeError("Something bad happened")
+        with MockDockerClient(mock_container):
+            safe_docker = SafeDockerClient()
+            assert safe_docker.run("image", "command") == 152
+
+    def test_volume_mount(self):
```
Comment (perdasilva) [inline ci/test_safe_docker_run.py]: Maybe this should be an internal method and the translation should be checked in the main function tests

File: ci/safe_docker_run.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+    parser.add_argument("--name",
+                        help="Assign a name to the container",
+                        default=None)
+
+    parser.add_argument("image", metavar="IMAGE")
+    parser.add_argument("command", metavar="COMMAND")
+    parser.add_argument("args", nargs='*', metavar="ARG")
+
+    args = parser.parse_args(args=command_line_arguments[1:])
```
Comment (perdasilva) [inline ci/safe_docker_run.py]: maybe pass in sys.argv[1:] instead of doing command_line_arguments[1:] here...

File: ci/docker_login.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                        help="Endpoint Url",
+                        type=str,
+                        default=os.environ.get("DOCKERHUB_SECRET_ENDPOINT_URL", None))
+
+    parser.add_argument("--secret-endpoint-region",
+                        help="AWS Region",
+                        type=str,
+                        default=os.environ.get("DOCKERHUB_SECRET_ENDPOINT_REGION", None))
+
+    args = parser.parse_args(args=sys_argv[1:])
```
Comment (perdasilva) [inline ci/docker_login.py]: Maybe pass in sys.argv[1:] to the main method, instead of doing sys_argv[1:] here

File: ci/docker_login.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+
+def main(sys_argv):
+    config_logging()
+
+    parser = argparse.ArgumentParser(
+        description="Safe docker login utility to avoid leaking passwords",
+        epilog=""
+    )
+    parser.add_argument("--secret-name",
```
Comment (larroy) [inline ci/docker_login.py:106]: Beautiful, if we could ask for the cherry on top of the cake, being able to pass user and password in case we don't want / we can't use secrets manager it would be great.

File: ci/safe_docker_run.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+            containers = self._docker_client.containers.list()
+            if containers:
+                logging.info("Other running containers: %s", [self._trim_container_id(x.id) for x in containers])
+        except NotFound as e:
+            logging.info("Container was stopped before cleanup started: %s", e)
+
+        return ret
+
+
+def volume_mount(volume_dfn: str) -> Dict[str, Any]:
```
Comment (larroy) [inline ci/safe_docker_run.py]: Nice!  for icing on the cake adding a docstring :-)

File: ci/test_safe_docker_run.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
+#!/usr/bin/env python3
```
Comment (larroy) [inline ci/test_safe_docker_run.py:1]: This is great

File: ci/util.py [hunk 11]
```diff
@@ -113,3 +116,14 @@ def chdir_to_script_directory():
     os.chdir(base)
 
 
+def script_name() -> str:
+    """:returns: script name with leading paths removed"""
+    return os.path.split(sys.argv[0])[1]
+
+
+def config_logging():
```
Comment (larroy) [inline ci/util.py:125]: Could we update this with something like:

```

def config_logging():
    logging.config.fileConfig(os.getenv('LOGGING_CONF', 'logging.conf'))
```


```
[loggers]
keys=root,simpleExample

[handlers]
keys=consoleHandler

[formatters]
keys=simpleFormatter

[logger_root]
level=DEBUG
handlers=consoleHandler

[logger_simpleExample]
level=DEBUG
handlers=consoleHandler
qualname=simpleExample
propagate=0

[handler_consoleHandler]
class=StreamHandler
level=DEBUG
formatter=simpleFormatter
args=(sys.stdout,)

[formatter_simpleFormatter]
format=%(asctime)s - %(name)s - %(levelname)s - %(message)s
datefmt=
```

I'm moving away from these hardcoded logging statements.

**Review by larroy — CHANGES_REQUESTED** [review]
Great work @perdasilva thanks a lot for this. 

Everything looks quite good in general, just the     jenkins_env_vars = ['BUILD_NUMBER', 'BUILD_ID', 'BUILD_TAG']   I don't see them set. these have to be set and percolated through the containers so Jenkins can kill runaway containers as the signal is not always sent / there's no time to act when the jobs are interrupted.

**Comment by larroy** [comment]
Also, great stuff with the tests and mocking, great to learn from you.

File: ci/docker_login.py (hunk 8 above)
Comment (perdasilva) [reply ci/docker_login.py:106]: I won't put this through because we had a whole CoE on leaked docker passwords - this is why we ended up having to use the secrets manager =P I'm sure if I add it, Marco will flip... ._.

**Comment by perdasilva** [comment]
@larroy thank you so much for taking the time to review - it was specially important to get the zombie guarding right - I will put through the the env vars thing through and add the tests. This way it is documented and ensured! Thanks again ^^

**Comment by larroy** [comment]
I will help review this again in the next couple of days. Feel free to ping me if I'm not responsibe as I missed your latest changes. Thanks again for following up on this and let's try to get it merged so we can open smaller PRs on top.

**Comment by perdasilva** [comment]
@larroy - after addressing your comments I created a PR against incubator - it's been merged sans issues thanks to you ^^ we are now producing nightly python images for mxnet XD
